src/reconcile/rec_visibility.c

```diff
@@ -265,64 +256,54 @@ static int
+    if (F_ISSET(upd_select->upd, WT_UPDATE_HS | WT_UPDATE_HS_MAX_STOP)) {
+        if (upd_select->tombstone == NULL)
+            tombstone = NULL;
+        else if (F_ISSET(upd_select->tombstone, WT_UPDATE_HS))
+            tombstone = upd_select->tombstone;
+        else {
+            tombstone = NULL;
+            WT_ASSERT(session, !F_ISSET(upd_select->tombstone, WT_UPDATE_HS_MAX_STOP));
         }
+        WT_RET(__rec_delete_hs_upd_save(session, r, ins, rip, upd_select->upd, tombstone));
+        return (0);
+    }
 
-        if (delete_upd->txnid == WT_TXN_ABORTED)
+    seen_committed = !WT_TIME_WINDOW_HAS_START_PREPARE(&upd_select->tw);
+
+    /* Delete all the history store record with a max stop timestamp. */
+    for (delete_upd = upd_select->upd->next; delete_upd != NULL; delete_upd = delete_upd->next) {
+        if ((txnid = delete_upd->txnid) == WT_TXN_ABORTED)
             continue;
 
-        if (F_ISSET(delete_upd, WT_UPDATE_TO_DELETE_FROM_HS)) {
-            WT_ASSERT_ALWAYS(session,
-              F_ISSET(delete_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_HS),
-              "Attempting to remove an update from the history store in WiredTiger, but the "
-              "update was missing.");
-            find_triggering_prepare = false;
-            if (triggering_prepare_upd != NULL) {
-                uint64_t txnid;
-                WT_ACQUIRE_READ(txnid, triggering_prepare_upd->txnid);
-                if (txnid == WT_TXN_ABORTED)
-                    txnid = triggering_prepare_upd->upd_saved_txnid;
-                /*
-                 * It must be the operation on this prepared update that triggered us to delete the
-                 * update from the history store. If it is the same prepared update we are writing
-                 * to disk, don't delete the update from the history store now. We need to wait the
-                 * prepared update to become stable.
-                 */
-                if (txnid == upd_select->tw.start_txn)
-                    break;
-            }
+        if (!seen_committed && upd_select->tw.start_txn != txnid)
+            seen_committed = true;
 
-            if (delete_upd->type == WT_UPDATE_TOMBSTONE)
-                delete_tombstone = delete_upd;
-            else {
-                WT_RET(
-                  __rec_delete_hs_upd_save(session, r, ins, rip, delete_upd, delete_tombstone));
+        if (F_ISSET(delete_upd, WT_UPDATE_HS_MAX_STOP)) {
+            WT_ASSERT(session, delete_upd->type != WT_UPDATE_TOMBSTONE);
+            /*
+             * Don't delete it from the history store if we again write a prepared update to disk
+             * and there is nothing in between the prepared update and the update with the max stop
+             * point in the history store.
+             */
+            if (!seen_committed)
                 break;
-            }
+            WT_RET(__rec_delete_hs_upd_save(session, r, ins, rip, delete_upd, NULL));
+            break;
```

Comment:
The comment says "Delete all the history store record with a max stop timestamp" but if we break here, only one record with max stop timestamp will be deleted?

I also have another question for the break in line `300`. Consider this scenario:
```
[upd_select (prepared)] -> [hs_entry1 (WT_UPDATE_HS_MAX_STOP)] -> [committed_txn] -> [hs_entry2 (WT_UPDATE_HS_MAX_STOP)]
```
Since `upd_select` is written as prepared, `seen_committed` is false and we will break early at `hs_entry1` and won't delete `hs_entry2`. Is this expected?

Reply:
This case is not possible. There has only one maximum record with max stop timestamp in the history store for a key.

Reply:
done

Reply:
why is that the case? If no reconciliation is called between two transactions, how is the previous record with max ts be deleted?

Reply:
Writing a prepared to disk needs a reconciliation. At that reconciliation, it will delete the previous max stop time and then insert the new one as max stop. So you cannot have two updates both with max in hs. Anyway you will get an error when you insert the second one with max stop when there is an existing one in hs.